### PR TITLE
Fix for resource#resolveDeferredsWhenWithinViewports_ to use parseFloat

### PR DESCRIPTION
**Pull request by keithwrightbos**
Fix for https://github.com/ampproject/amphtml/issues/16218

### CODE REVIEW

File: src/service/resource.js [hunk 1]
```diff
@@ -695,10 +695,10 @@ export class Resource {
     const viewportRatio = this.getDistanceViewportRatio();
     const keys = Object.keys(this.withViewportDeferreds_);
     for (let i = 0; i < keys.length; i++) {
-      const viewportInt = dev().assertNumber(parseInt(keys[i], 10));
-      if (this.isWithinViewportRatio(viewportInt, viewportRatio)) {
-        this.withViewportDeferreds_[viewportInt].resolve();
-        delete this.withViewportDeferreds_[viewportInt];
+      const viewportNum = dev().assertNumber(parseFloat(keys[i], 10));
+      if (this.isWithinViewportRatio(viewportNum, viewportRatio)) {
+        this.withViewportDeferreds_[viewportNum].resolve();
```
Comment (zhouyx) [inline src/service/resource.js]: I thought the fix was to check if `this.withViewportDeferreds_[viewportNum]` is defined.

File: src/service/resource.js [hunk 2]
```diff
@@ -695,10 +695,10 @@ export class Resource {
     const viewportRatio = this.getDistanceViewportRatio();
     const keys = Object.keys(this.withViewportDeferreds_);
     for (let i = 0; i < keys.length; i++) {
-      const viewportInt = dev().assertNumber(parseInt(keys[i], 10));
-      if (this.isWithinViewportRatio(viewportInt, viewportRatio)) {
-        this.withViewportDeferreds_[viewportInt].resolve();
-        delete this.withViewportDeferreds_[viewportInt];
+      const viewportNum = dev().assertNumber(parseFloat(keys[i], 10));
```
Comment (zhouyx) [inline src/service/resource.js]: parseFloat(1.25, 10) = 1.25, which leads to `this.withViewportDeferreds_[1.25]`. I think we should keep using `parseInt`

Comment (zhouyx) [reply src/service/resource.js]: nvm, just learnt that the key here is a string, where we are expecting this.withViewportDeferreds_['1.25']

File: src/service/resource.js [hunk 3]
```diff
@@ -695,10 +695,13 @@ export class Resource {
...
-      const viewportInt = dev().assertNumber(parseInt(keys[i], 10));
-      if (this.isWithinViewportRatio(viewportInt, viewportRatio)) {
-        this.withViewportDeferreds_[viewportInt].resolve();
-        delete this.withViewportDeferreds_[viewportInt];
+      // Despite type indicating object keys are numbers, JS stores them as
+      // strings so they need to be coverted back to floats in order to be
+      // used as numbers in isWithinViewportRatio.
+      const viewportNum = dev().assertNumber(parseFloat(keys[i]));
+      if (this.isWithinViewportRatio(viewportNum, viewportRatio)) {
+        this.withViewportDeferreds_[viewportNum].resolve();
```
Comment (jridgewell) [inline src/service/resource.js]: Use `keys[i]`

Comment (keithwrightbos) [reply src/service/resource.js]: When I do so, I get ERROR - restricted index type
found   : string
required: number

Casting via /** @type {number} */(keys[i]) doesn't help

**Pull request merged**